File: Algorithm/Ford-Fulkerson/Edmonds-Karp/atcoder/A68/A68.php
<?php

fscanf(STDIN, "%d %d", $N, $M);

// グラフ初期化
$capacity = array_fill(0, $N + 1, array_fill(0, $N + 1, 0));
$graph = array_fill(0, $N + 1, []);

for ($i = 0; $i < $M; $i++) {
    fscanf(STDIN, "%d %d %d", $a, $b, $c);
    $capacity[$a][$b] += $c; // 同じ辺が複数ある場合は加算
    $graph[$a][] = $b;
    $graph[$b][] = $a; // 逆辺も追加（残余グラフ用）
}

// BFSで増加経路を探索
function bfs($s, $t, &$parent, $capacity, $graph)
{
    $visited = array_fill(0, count($graph), false);
    $queue = [$s];
    $visited[$s] = true;
    $parent[$s] = -1;

    while (!empty($queue)) {
        $u = array_shift($queue);
        foreach ($graph[$u] as $v) {
            if (!$visited[$v] && $capacity[$u][$v] > 0) {
                $parent[$v] = $u;
                $visited[$v] = true;
                if ($v == $t) return true;
                $queue[] = $v;
            }
        }
    }

    return false;
}

// 最大流計算（Edmonds-Karp）
function maxFlow($s, $t, &$capacity, $graph)
{
    $flow = 0;
    $parent = [];

    while (bfs($s, $t, $parent, $capacity, $graph)) {
        $pathFlow = PHP_INT_MAX;
        for ($v = $t; $v != $s; $v = $parent[$v]) {
            $u = $parent[$v];
            $pathFlow = min($pathFlow, $capacity[$u][$v]);
        }

        for ($v = $t; $v != $s; $v = $parent[$v]) {
            $u = $parent[$v];
            $capacity[$u][$v] -= $pathFlow;
            $capacity[$v][$u] += $pathFlow;
        }

        $flow += $pathFlow;
    }

    return $flow;
}

echo maxFlow(1, $N, $capacity, $graph) . PHP_EOL;

File: Algorithm/Ford-Fulkerson/Edmonds-Karp/atcoder/A68/A68.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const INF = 1 << 30

func main() {
	reader := bufio.NewReader(os.Stdin)

	// 1行目: N, M
	line, _ := reader.ReadString('\n')
	parts := strings.Fields(line)
	N, _ := strconv.Atoi(parts[0])
	M, _ := strconv.Atoi(parts[1])

	// グラフ構築
	capacity := make([][]int, N+1)
	for i := range capacity {
		capacity[i] = make([]int, N+1)
	}

	graph := make([][]int, N+1)

	for i := 0; i < M; i++ {
		line, _ := reader.ReadString('\n')
		parts := strings.Fields(line)
		a, _ := strconv.Atoi(parts[0])
		b, _ := strconv.Atoi(parts[1])
		c, _ := strconv.Atoi(parts[2])

		capacity[a][b] += c
		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a) // 逆辺も追加（残余グラフ）
	}

	fmt.Println(maxFlow(1, N, capacity, graph))
}

func bfs(s, t int, parent []int, capacity [][]int, graph [][]int) bool {
	N := len(capacity) - 1
	visited := make([]bool, N+1)
	queue := []int{s}
	visited[s] = true
	parent[s] = -1

	for len(queue) > 0 {
		u := queue[0]
		queue = queue[1:]

		for _, v := range graph[u] {
			if !visited[v] && capacity[u][v] > 0 {
				parent[v] = u
				visited[v] = true
				if v == t {
					return true
				}
				queue = append(queue, v)
			}
		}
	}
	return false
}

func maxFlow(s, t int, capacity [][]int, graph [][]int) int {
	flow := 0
	parent := make([]int, len(capacity))

	for bfs(s, t, parent, capacity, graph) {
		pathFlow := INF
		v := t
		for v != s {
			u := parent[v]
			if capacity[u][v] < pathFlow {
				pathFlow = capacity[u][v]
			}
			v = u
		}

		v = t
		for v != s {
			u := parent[v]
			capacity[u][v] -= pathFlow
			capacity[v][u] += pathFlow
			v = u
		}

		flow += pathFlow
	}
	return flow
}
